File: leetcode/search-a-2d-matrix/search-a-2d-matrix.cpp
/**************************************
* @author dujiong
* @date 2016.11.29
* @version V0.1
**************************************/

class Solution {
public:
    bool searchMatrix(vector<vector<int> > &matrix, int target) {
        if(matrix.size()==0){
            return false;
        }
        int n=matrix.size();
        int m=matrix[0].size();
      
        int index;
		for(int i=0;i<n;i++){
            index = std::lower_bound(matrix[i].begin(), matrix[i].end(), target)-matrix[i].begin();
            
            if(index<m && matrix[i][index]==target){
                return true;
            } 
                
        }
        return false;
    }
};

File: leetcode-golang/Search a 2D Matrix/Search a 2D Matrix.go
//核心，从右上角开始剥洋葱
func searchMatrix(matrix [][]int, target int) bool {
	if len(matrix) == 0 {
		return false
	}
	row, column := 0, len(matrix[0])-1
	for row < len(matrix) && column >= 0 {
		if matrix[row][column] == target {
			return true
		} else if matrix[row][column] < target {
			row++
		} else {
			column--
		}
	}
	return false
}
